_posts/2018-01-24-kubernetes-and-hokusai.markdown

```diff
@@ -0,0 +1,134 @@
+
+<!-- more -->
+
+While Heroku is easy to get started with, costs scale up quickly. And, as described in our [2017 tech stack post][stack], our team is moving more and more towards [Kubernetes][]. I had almost no experience with Kubernetes before last week, and I was intimidated by the Kubernetes web UI. With some help from my colleague Isac, who wrote the [Hokusai][] tool, I was able to get a staging environment up and running in under a day.
+
+But let's step back first.
+
+My background is in iOS software development, so spinning up new servers isn't something I do often. When I _do_, I usually use Heroku. After deploying to it, it feels like Kubernetes is a kind of hosted Heroku: it handles scaling up instances, managing worker/db/other instances, load-balancers, environment variables, promoting from staging to production – all that stuff. But Kubernetes' sophistication comes with a sophisticated user interface. 
+
+So basically, Kubernetes is to Heroku what Hokusai is to the Heroku command-line tool.
+
+Hokusai provides [a bunch of commands][commands] for interacting with the Kubernetes cluster. Deploying my new Rails app to Kubernetes involved a few steps, but most of the work was handled automatically by Hokusai.
+
+First, I installed and setup Hokusai locally (with required environment variables for AWS access). I then ran the following command to scaffold out everything.
+
+```sh
+hokusai setup --aws-account-id ARTSY_ACCOUNT_ID --project-type ruby-rails
+```
+
+In addition to staging- and production-specific config files, this command creates a `Dockerfile`. See, where Heroku uses Dynos as a high level of abstraction, Kubernetes uses [Docker][] images (as a slightly less high a level of abstraction). Docker is a technology I'm familiar with, and I managed to configure the generated `Dockerfile` and `hokusai/*.yml` config files pretty quickly. At this point, I could run `hokusai dev start` to start a development Docker container, or `hokusai test` to run RSpec tests. Nothing fancy yet, but that verifies that everything is working so far.
+
+Next up was to use Hokusai in our CI environment. [Circle CI 2.0][circle] is very Docker-oriented, so we set up everything using their [Workflows][]. This is a much higher level of abstraction for CI configuration than I'm used to, but I got the hang of it quickly. I created a job to run RSpec tests through Hokusai, a job to run [Danger][], a job to build and push a Docker image to our S3 bucket, and a job to deploy that image to the Kubernetes cluster. Finally, I added the workflows to build and deploy automatically after successful builds on the `master` branch.
+
+Here's a complete, redacted copy of our Circle config:
```

Comment:
Is "complete, redacted" contradictory? :) Maybe "redacted, but working"?

Reply:
Mmm yeah.